File: lstgen/generators/php/BigDecimal.php
use Brick\Math\BigDecimal as BaseBigDecimal;
use Brick\Math\RoundingMode as RoundingMode;

/**
 * A proxy class to use BigDecimal methods
 * as they are found in PAP XML.
 * Requires brick/math >= 0.6.0
 */
class BigDecimal {

    const ROUND_DOWN = RoundingMode::DOWN;
    const ROUND_UP = RoundingMode::UP;

    private $bd;

    public function __construct($value) {
        $this->bd = BaseBigDecimal::of($value);
    }

    public function divide($other, $scale=null, $rounding=null) {
        if (!$scale && !$rounding) {
            return new BigDecimal($this->bd->exactlyDividedBy($other));
        }
        return new BigDecimal($this->bd->dividedBy($other, $scale, $rounding));
    }

    public static function valueOf($value) {
        return new BigDecimal($value);
    }

    public function multiply($other) {
        return new BigDecimal($this->bd->multipliedBy($other));
    }

    public function setScale($scale, $rounding) {
        return new BigDecimal($this->bd->toScale($scale, $rounding));
    }

    public function add($other) {
        return new BigDecimal($this->bd->plus($other));
    }

    public function subtract($other) {
        return new BigDecimal($this->bd->minus($other));
    }

    public function longValue() {
        return $this->bd->toInt();
    }

    public function floatValue() {
        return $this->bd->toFloat();
    }

    public function compareTo($other) {
        return $this->bd->compareTo($other);
    }

    public function __toString() {
        return $this->bd->__toString();
    }

    public static function zero() {
        return BigDecimal::valueOf(0);
    }

    public static function one() {
        return BigDecimal::valueOf(1);
    }

    public static function ten() {
        return BigDecimal::valueOf(10);
    }

}

File: lstgen/generators/golang/BigDecimal.go
import "math"

const (
	ROUND_DOWN = 1
	ROUND_UP   = 2
)

type BigDecimal struct {
	V float64
}

func (d BigDecimal) Float64() float64 {
	return d.V
}

func (d BigDecimal) Divide(other BigDecimal, roundParams ...int) BigDecimal {
	scale := 0
	rounding := 0
	if len(roundParams) == 2 {
		scale = roundParams[0]
		rounding = roundParams[1]
	} else {
		return BigDecimal{d.V / other.V}
	}

	return BigDecimal{d.V / other.V}.SetScale(scale, rounding)
}

func (d BigDecimal) ValueOf(value float64) BigDecimal {
	return BigDecimal{value}
}

func (d BigDecimal) Multiply(other BigDecimal) BigDecimal {
	return BigDecimal{d.V * other.V}
}

func (d BigDecimal) Add(other BigDecimal) BigDecimal {
	return BigDecimal{d.V + other.V}
}

func (d BigDecimal) Subtract(other BigDecimal) BigDecimal {
	return BigDecimal{d.V - other.V}
}

func (d BigDecimal) IntPart() int64 {
	i, _ := math.Modf(d.V)
	return int64(i)
}

func (d BigDecimal) CompareTo(other BigDecimal) int {
	if d.V > other.V {
		return 1
	} else if d.V < other.V {
		return -1
	}
	return 0
}

func (d BigDecimal) SetScale(scale int, rounding int) BigDecimal {
	exp := math.Pow(10, float64(scale))

	var v float64
	if rounding == ROUND_DOWN {
		if d.V < 0 {
			v = math.Ceil(d.V * exp)
		} else {
			v = math.Floor(d.V * exp)
		}
	} else if rounding == ROUND_UP {
		if d.V < 0 {
			v = math.Floor(d.V * exp)
		} else {
			v = math.Ceil(d.V * exp)
		}
	} else {
		v = math.Round(d.V * exp)
	}
	v = v / exp
	return BigDecimal{v}
}

func NewFromInt(value int64) BigDecimal {
	return BigDecimal{float64(value)}
}

func NewFromFloat(value float64) BigDecimal {
	return BigDecimal{value}
}
